packages/primitives/popover/src/popover-root.directive.ts

```diff
@@ -0,0 +1,157 @@
+import { DOCUMENT } from '@angular/common';
+import {
+    computed,
+    contentChild,
+    Directive,
+    effect,
+    forwardRef,
+    inject,
+    InjectionToken,
+    input,
+    OnInit,
+    output,
+    signal,
+    untracked,
+    ViewContainerRef
+} from '@angular/core';
+import { RdxPopoverContentDirective } from './popover-content.directive';
+import { RdxPopoverTriggerDirective } from './popover-trigger.directive';
+import { RdxPopoverState } from './popover.types';
+
+export const RdxPopoverRootToken = new InjectionToken<RdxPopoverRootDirective>('RdxPopoverRootToken');
+
+export function injectPopoverRoot(): RdxPopoverRootDirective {
+    return inject(RdxPopoverRootToken);
+}
+
+@Directive({
+    selector: '[rdxPopoverRoot]',
+    standalone: true,
+    providers: [
+        {
+            provide: RdxPopoverRootToken,
+            useExisting: forwardRef(() => RdxPopoverRootDirective)
+        }
+    ],
+    exportAs: 'rdxPopoverRoot'
+})
+export class RdxPopoverRootDirective implements OnInit {
+    /**
+     * The open state of the popover when it is initially rendered. Use when you do not need to control its open state.
+     */
+    readonly defaultOpen = input<boolean>(false);
+    /**
+     * The controlled open state of the popover. Must be used in conjunction with onOpenChange.
+     */
+    readonly open = input<boolean | undefined>();
+
+    /**
+     * Event handler called when the open state of the popover changes.
+     */
+    readonly onOpenChange = output<boolean>();
+
+    /** @ignore */
+    readonly popoverContentDirective = contentChild.required(RdxPopoverContentDirective);
```

Comment:
we have some collisions (tooltip also).

Ensure that rdxTooltipRoot properly associates rdxTooltipTrigger and rdxTooltipContent within the same container.

Issues may arise if @ContentChild or @ContentChildren selects elements globally instead of limiting the scope to the nearest rdxTooltipRoot.

How reproduce:

1. Create several popovers on one page
2. set `sideOffset` on first popover
```
<ng-template [sideOffset]="29" rdxTooltipContent>
```
3. `sideOffset` will be applied on the last

Reply:
> we have some collisions (tooltip also).
> 
> Ensure that rdxTooltipRoot properly associates rdxTooltipTrigger and rdxTooltipContent within the same container.
> 
> Issues may arise if @ContentChild or @ContentChildren selects elements globally instead of limiting the scope to the nearest rdxTooltipRoot.
> 
> How reproduce:
> 
>     1. Create several popovers on one page
> 
>     2. set `sideOffset` on first popover
> 
> 
> ```
> <ng-template [sideOffset]="29" rdxTooltipContent>
> ```
> 
>     3. `sideOffset` will be applied on the last

Hmmm I will check it

Reply:
yes, some issue is there. I apply `sideOffset` only to the first one but all of them get it. I need to investigate it more.

![obraz](https://github.com/user-attachments/assets/f5734f46-2dd4-483b-8a75-e13a3fd3548b)

Reply:
Fixed :) multiple popovers on one page work perfect ;)

![obraz](https://github.com/user-attachments/assets/7cd41019-5982-435b-8e90-c21bd9b4c5c5)

Reply:
@pimenovoleg I cannot see the issue since `contentChild` and `contentChildren` are limited to the content of an element that the directive or component is attached to.

But I might not understad what you are saying :) Could you elaborate bit more on it? :)

Reply:
update this for fix:

```typescript
export function injectPopoverRoot(): RdxPopoverRootDirective {
    assertInInjectionContext(injectPopoverRoot);
    return <RdxPopoverRootDirective>inject(RdxPopoverRootToken, { host: true, optional: true });
}
```

Adding `@Host()` and `@Optional()` to the `RdxPopoverContentDirective` resolves the issue by limiting the dependency scope.

Why does this work?

`@Host()`: Ensures Angular looks for the dependency (`RdxPopoverRootComponent`) only in the closest parent, rather than searching the entire DI hierarchy. This prevents accidental use of the wrong `RdxPopoverRootComponent` instance when multiple such constructs are present on the page.
`@Optional()`: Avoids errors when the parent component is missing and allows us to throw a custom error with a clear message.

Result: Each `rdxPopoverContent` directive is now correctly linked to its own `RdxPopoverRoot`, fixing issues with sideOffset and other parameters.

Reply:
sry, I dont see r changes. That works perfectly 👍

Reply:
I am not really sure it is a problem. I cannot see any issue of that kind in `popover` so far.
I've added `multiple` case to the stories and they work fine.

Yeah, `inject` works exactly the way you are descibing.

Injectors in Angular are hierarchical so there is no room for such an issue unless there is a popover inside a popover.
This is how I see it. And I see no issues that should get fixed :)

I am quite sure on what I am saying but you know I can always be wrong :)

https://github.com/user-attachments/assets/40bcb275-473c-476e-a9e0-1e3df25bab43